crates/circuit/src/dag_circuit.rs

```diff
@@ -0,0 +1,6474 @@
+                .map_err(|e| match e {
+                    rustworkx_core::dag_algo::TopologicalSortError::CycleOrBadInitialState => {
+                        PyValueError::new_err(format!("{}", e))
+                    }
+                    rustworkx_core::dag_algo::TopologicalSortError::KeyError(_) => {
+                        unreachable!()
+                    }
+                })?;
+        Ok(nodes.into_iter())
+    }
+
+    fn topological_op_nodes(&self) -> PyResult<impl Iterator<Item = NodeIndex> + '_> {
+        Ok(self.topological_nodes()?.filter(|node: &NodeIndex| {
+            matches!(self.dag.node_weight(*node), Some(NodeType::Operation(_)))
+        }))
+    }
+
+    fn topological_key_sort(
+        &self,
+        py: Python,
+        key: &Bound<PyAny>,
+    ) -> PyResult<impl Iterator<Item = NodeIndex>> {
+        // This path (user provided key func) is not ideal, since we no longer
+        // use a string key after moving to Rust, in favor of using a tuple
+        // of the qargs and cargs interner IDs of the node.
+        let key = |node: NodeIndex| -> PyResult<String> {
+            let node = self.get_node(py, node)?;
+            key.call1((node,))?.extract()
+        };
+        Ok(
+            rustworkx_core::dag_algo::lexicographical_topological_sort(&self.dag, key, false, None)
+                .map_err(|e| match e {
+                    rustworkx_core::dag_algo::TopologicalSortError::CycleOrBadInitialState => {
+                        PyValueError::new_err(format!("{}", e))
+                    }
+                    rustworkx_core::dag_algo::TopologicalSortError::KeyError(ref e) => {
+                        e.clone_ref(py)
+                    }
+                })?
+                .into_iter(),
+        )
+    }
+
+    fn is_wire_idle(&self, wire: &Wire) -> PyResult<bool> {
+        let (input_node, output_node) = match wire {
+            Wire::Qubit(qubit) => (self.qubit_input_map[qubit], self.qubit_output_map[qubit]),
+            Wire::Clbit(clbit) => (self.clbit_input_map[clbit], self.clbit_output_map[clbit]),
+            Wire::Var(var) => (
+                self.var_input_map.get(var).unwrap(),
+                self.var_output_map.get(var).unwrap(),
+            ),
+        };
+
+        let child = self
+            .dag
+            .neighbors_directed(input_node, Outgoing)
+            .next()
+            .ok_or_else(|| {
+                DAGCircuitError::new_err(format!(
+                    "Invalid dagcircuit input node {:?} has no output",
+                    input_node
+                ))
+            })?;
+
+        Ok(child == output_node)
+    }
+
+    fn may_have_additional_wires(&self, py: Python, instr: &PackedInstruction) -> bool {
+        let has_condition = match instr.condition() {
+            None => false,
+            Some(condition) => !condition.bind(py).is_none(),
+        };
```

Comment:
If we're ever storing an explicit `py.None()` in a `Some(condition)`, that probably indicates a logic bug elsewhere.

Reply:
Removed this additional check in 4d68875717 - at worst it fails safe anyway.